File: ui/src/data/sorting.ts
import type { PropertyOpportunity, SortConfig, SortField, Status, Confidence } from '../types/property'

/**
 * Status priority order (VIABLE first per PRODUCT.md ranking rules)
 */
const STATUS_ORDER: Record<Status, number> = {
  VIABLE: 0,
  WATCHLIST: 1,
  REJECTED: 2,
}

/**
 * Confidence priority order (HIGH first per PRODUCT.md ranking rules)
 */
const CONFIDENCE_ORDER: Record<Confidence, number> = {
  HIGH: 0,
  MEDIUM: 1,
  LOW: 2,
}

/**
 * Compare two opportunities by a given field (ascending order)
 * All comparisons return positive if a > b, negative if a < b
 */
function compareByField(a: PropertyOpportunity, b: PropertyOpportunity, field: SortField): number {
  switch (field) {
    case 'status':
      return STATUS_ORDER[a.status] - STATUS_ORDER[b.status]

    case 'confidence':
      return CONFIDENCE_ORDER[a.confidence] - CONFIDENCE_ORDER[b.confidence]

    case 'capRate':
      return a.capRate - b.capRate

    case 'noi':
      return a.noi - b.noi

    default:
      return 0
  }
}

/**
 * Sort opportunities according to PRODUCT.md ranking rules:
 * 1. Status (VIABLE first)
 * 2. Confidence in the analysis
 * 3. Cap rate (higher is better)
 * 4. Estimated annual cash generation (NOI, higher is better)
 */
export function sortOpportunities(
  opportunities: PropertyOpportunity[],
  config?: SortConfig
): PropertyOpportunity[] {
  const sorted = [...opportunities]

  if (config) {
    sorted.sort((a, b) => {
      const result = compareByField(a, b, config.field)
      return config.direction === 'desc' ? -result : result
    })
  } else {
    sorted.sort((a, b) => {
      let result = compareByField(a, b, 'status')
      if (result !== 0) return result

      result = compareByField(a, b, 'confidence')
      if (result !== 0) return result

      result = -compareByField(a, b, 'capRate')
      if (result !== 0) return result

      return -compareByField(a, b, 'noi')
    })
  }

  return sorted
}

/**
 * Get the next sort direction for toggling
 */
export function getNextSortDirection(
  currentConfig: SortConfig | undefined,
  field: SortField
): SortConfig {
  if (currentConfig?.field === field) {
    return {
      field,
      direction: currentConfig.direction === 'asc' ? 'desc' : 'asc',
    }
  }
  const defaultDesc: SortField[] = ['capRate', 'noi']
  return {
    field,
    direction: defaultDesc.includes(field) ? 'desc' : 'asc',
  }
}

File: streamlit/sorting.py
"""Sort opportunities per PRODUCT.md ranking rules (parity with ui/src/data/sorting.ts)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Optional

from db_client.types import PropertyOpportunity, ConfidenceLevel, PropertyStatus

SortField = Literal['status', 'confidence', 'cap_rate', 'noi']
SortDirection = Literal['asc', 'desc']

STATUS_ORDER: dict[PropertyStatus, int] = {
    'VIABLE': 0,
    'WATCHLIST': 1,
    'REJECTED': 2,
}

CONFIDENCE_ORDER: dict[ConfidenceLevel, int] = {
    'HIGH': 0,
    'MEDIUM': 1,
    'LOW': 2,
}


@dataclass
class SortConfig:
    field: SortField
    direction: SortDirection


def _compare_by_field(a: PropertyOpportunity, b: PropertyOpportunity, field: SortField) -> float:
    if field == 'status':
        return STATUS_ORDER[a.status] - STATUS_ORDER[b.status]
    if field == 'confidence':
        return CONFIDENCE_ORDER[a.confidence] - CONFIDENCE_ORDER[b.confidence]
    if field == 'cap_rate':
        return a.cap_rate - b.cap_rate
    if field == 'noi':
        return a.noi - b.noi
    return 0.0


def sort_opportunities(
    opportunities: list[PropertyOpportunity],
    config: Optional[SortConfig] = None,
) -> list[PropertyOpportunity]:
    """Sort opportunities; default order matches React UI."""
    sorted_list = list(opportunities)

    if config:
        sorted_list.sort(
            key=lambda item: _field_sort_key(item, config.field),
            reverse=config.direction == 'desc',
        )
        return sorted_list

    sorted_list.sort(
        key=lambda o: (
            STATUS_ORDER[o.status],
            CONFIDENCE_ORDER[o.confidence],
            -o.cap_rate,
            -o.noi,
        )
    )
    return sorted_list


def _field_sort_key(item: PropertyOpportunity, field: SortField) -> float | int:
    if field == 'status':
        return STATUS_ORDER[item.status]
    if field == 'confidence':
        return CONFIDENCE_ORDER[item.confidence]
    if field == 'cap_rate':
        return item.cap_rate
    return item.noi


def get_next_sort_direction(
    current: Optional[SortConfig],
    field: SortField,
) -> SortConfig:
    """Toggle sort direction; matches ui/src/data/sorting.ts."""
    if current and current.field == field:
        return SortConfig(field=field, direction='desc' if current.direction == 'asc' else 'asc')

    default_desc: set[SortField] = {'cap_rate', 'noi'}
    return SortConfig(field=field, direction='desc' if field in default_desc else 'asc')
